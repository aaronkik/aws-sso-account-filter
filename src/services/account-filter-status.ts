import { ChromeStorageSync } from '../repositories';

const accountFiltersStatusKey = 'accountFilterStatus';

type GetAccountFilterStatus = {
  [accountFiltersStatusKey]?: boolean;
};

class AccountFilterStatus {
  #storageKey = accountFiltersStatusKey;
  #storage = new ChromeStorageSync();

  async #set(status: boolean) {
    return this.#storage.set({ accountFilterStatus: status });
  }

  async get() {
    return new Promise<boolean>((resolve, reject) => {
      this.#storage
        .get<GetAccountFilterStatus>(this.#storageKey)
        .then((result) => {
          if (
            result &&
            'accountFilterStatus' in result &&
            typeof result.accountFilterStatus === 'boolean'
          ) {
            resolve(result.accountFilterStatus);
          } else {
            this.enable();
            resolve(true);
          }
        })
        .catch((error) => {
          reject(error);
        });
    });
  }

  async enable() {
    return this.#set(true);
  }

  async disable() {
    return this.#set(false);
  }
}

export const accountFilterStatus = new AccountFilterStatus();
