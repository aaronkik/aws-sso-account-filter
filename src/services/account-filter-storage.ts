import { ChromeStorageSync } from '~/repositories';

const accountFiltersStorageKey = 'accountFilters';

type AccountFilter = {
  id: string;
  filter: string;
};

type GetAccountFilterStorage = {
  [accountFiltersStorageKey]?: Array<AccountFilter>;
};

export class AccountFilterStorage {
  #storageKey = accountFiltersStorageKey;
  #storage = new ChromeStorageSync();

  async get() {
    return this.#storage.get<GetAccountFilterStorage>(this.#storageKey);
  }

  async set(accountFilters: Array<AccountFilter>) {
    return this.#storage.set({ accountFilters });
  }
}
