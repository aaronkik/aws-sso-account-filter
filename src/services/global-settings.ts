import { ChromeStorageSync } from '../repositories';

interface GetGlobalSettings {
  accountFilterStatus?: boolean;
  autoRefreshPage?: boolean;
}

export type GlobalSettingKey = keyof GetGlobalSettings;

class GlobalSettings extends ChromeStorageSync {
  async getValue(key: GlobalSettingKey): Promise<boolean> {
    const result = await super.get<GetGlobalSettings>(key);

    if (result && key in result && typeof result[key] === 'boolean') {
      return result[key];
    }

    await super.set({ [key]: true });
    return true;
  }

  async setValue(key: GlobalSettingKey, value: boolean): Promise<void> {
    return super.set({ [key]: value });
  }
}

export const globalSettings = new GlobalSettings();
