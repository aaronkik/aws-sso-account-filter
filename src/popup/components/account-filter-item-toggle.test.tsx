import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import type { ComponentPropsWithoutRef } from 'react';
import { describe, expect, test, vi } from 'vitest';
import { accountFilterStorage } from '../../services/account-filter-storage';
import AccountFilterItemToggle from './account-filter-item-toggle';

type AccountFilterItemToggleProps = ComponentPropsWithoutRef<typeof AccountFilterItemToggle>;

const defaultAccountFilterItemToggleProps: AccountFilterItemToggleProps = {
  accountFilterItem: {
    id: 'random-id',
    enabled: true,
    filter: 'FizzBuzz',
  },
};

const initialSetup = (props?: AccountFilterItemToggleProps) => {
  const mergedProps: AccountFilterItemToggleProps = {
    ...defaultAccountFilterItemToggleProps,
    ...props,
  };

  const user = userEvent.setup();
  const component = render(<AccountFilterItemToggle {...mergedProps} />);

  const switchButton = screen.getByRole('switch', {
    name: `${mergedProps.accountFilterItem.enabled ? 'Disable' : 'Enable'} ${
      mergedProps.accountFilterItem.filter
    } filter`,
  });

  return { component, switchButton, user };
};

describe('AccountFilterItemToggle', () => {
  test('matches snapshot when filter is enabled', () => {
    const {
      component: { container },
    } = initialSetup();

    expect(container).toMatchSnapshot();
  });

  test('matches snapshot when filter is disabled', () => {
    const props = {
      ...defaultAccountFilterItemToggleProps,
      accountFilterItem: {
        ...defaultAccountFilterItemToggleProps.accountFilterItem,
        enabled: false,
      },
    };

    const {
      component: { container },
    } = initialSetup(props);

    expect(container).toMatchSnapshot();
  });

  test(`GIVEN a filter item is enabled
        THEN expect the toggle button to be checked`, () => {
    const { switchButton } = initialSetup();

    expect(switchButton).toBeChecked();
  });

  test(`GIVEN a filter item is disabled
        THEN expect the toggle button to not be checked`, () => {
    const props = {
      ...defaultAccountFilterItemToggleProps,
      accountFilterItem: {
        ...defaultAccountFilterItemToggleProps.accountFilterItem,
        enabled: false,
      },
    };

    const { switchButton } = initialSetup(props);

    expect(switchButton).not.toBeChecked();
  });

  test(`GIVEN a filter item is enabled
        WHEN the toggle button is clicked
        THEN expect disabled method to be called on chrome storage`, async () => {
    const accountFilterStorageDisableSpy = vi.spyOn(accountFilterStorage, 'disable');

    const { switchButton, user } = initialSetup();

    expect(switchButton).toBeChecked();
    await user.click(switchButton);

    expect(accountFilterStorageDisableSpy).toHaveBeenNthCalledWith(1, {
      accountFilterId: defaultAccountFilterItemToggleProps.accountFilterItem.id,
    });
  });

  test(`GIVEN a filter item is disabled
        WHEN the toggle button is clicked
        THEN expect enable method to be called on chrome storage`, async () => {
    const props = {
      ...defaultAccountFilterItemToggleProps,
      accountFilterItem: {
        ...defaultAccountFilterItemToggleProps.accountFilterItem,
        enabled: false,
      },
    };

    const accountFilterStorageEnableSpy = vi.spyOn(accountFilterStorage, 'enable');

    const { switchButton, user } = initialSetup(props);

    expect(switchButton).not.toBeChecked();
    await user.click(switchButton);

    expect(accountFilterStorageEnableSpy).toHaveBeenNthCalledWith(1, {
      accountFilterId: props.accountFilterItem.id,
    });
  });
});
