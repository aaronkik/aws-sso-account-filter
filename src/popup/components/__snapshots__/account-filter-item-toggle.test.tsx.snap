// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AccountFilterItemToggle > matches snapshot when filter is disabled 1`] = `
<div>
  <button
    aria-checked="false"
    class="bg-slate-400 relative inline-flex h-7 w-11 min-w-11 items-center rounded-sm"
    data-headlessui-state=""
    id="headlessui-switch-«r1»"
    role="switch"
    tabindex="0"
    type="button"
  >
    <span
      class="sr-only"
    >
      Enable FizzBuzz filter
    </span>
    <span
      class="translate-x-1 inline-block h-5 w-4 transform rounded-sm bg-white transition"
    />
  </button>
</div>
`;

exports[`AccountFilterItemToggle > matches snapshot when filter is enabled 1`] = `
<div>
  <button
    aria-checked="true"
    class="bg-orange-400 relative inline-flex h-7 w-11 min-w-11 items-center rounded-sm"
    data-checked=""
    data-headlessui-state="checked"
    id="headlessui-switch-«r0»"
    role="switch"
    tabindex="0"
    type="button"
  >
    <span
      class="sr-only"
    >
      Disable FizzBuzz filter
    </span>
    <span
      class="translate-x-6 inline-block h-5 w-4 transform rounded-sm bg-white transition"
    />
  </button>
</div>
`;
